library/core/src/iter/adapters/zip.rs

```diff
@@ -471,10 +514,21 @@ impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B
 ///   `self.__iterator_get_unchecked(idx)`, assuming that the required traits are implemented.
 /// * It must also be safe to drop `self` after calling `self.__iterator_get_unchecked(idx)`.
 /// * If `T` is a subtype of `Self`, then it must be safe to coerce `self` to `T`.
+//
+// FIXME: Clarify interaction with SourceIter/InPlaceIterable. Calling `SouceIter::as_inner`
+// after `__iterator_get_unchecked` is supposed to be allowed.
+#[doc(hidden)]
+#[unstable(feature = "trusted_random_access", issue = "none")]
+#[rustc_specialization_trait]
+pub unsafe trait TrustedRandomAccess: TrustedRandomAccessNoCoerce {}
+
+/// Like [`TrustedRandomAccess`] but without any of the requirements / guarantees around
+/// coercions to subtypes after `__iterator_get_unchecked` (they aren’t allowed here!), and
+/// without the requirement that subtypes / supertypes implement [`TrustedRandomAccessNoCoerce`].
```

Comment:
Should this or its subtrait have some comment that the distinction only exists as a workaround an unintended loophole in `min_specialization` and it should be revisited once that is either fixed or full specialization becomes available?

Reply:
I don’t know what you’d mean by “loophole in `min_specialization`”. I guess it might make sense to state that the `TrustedRandomAccess` trait structure is subject to change and perhaps one could even link relevant issues?

Reply:
Well, `min_specialization` was supposed to prevent lifetime-dependent specialization. Although maybe this isn't quit the same problem they tried to address there (typechecking and monomorphization disagreeing on which method should be used) since the problem with Zip isn't that the wrong method is called, rather that we can switch between specializations between two method calls.

Reply:
>  perhaps one could even link relevant issues?

The examples there definitely helped me understand the problem, yes.

Reply:
I personally came to feel like the fact that subtype coercions using variance of adapters such as `Zip` can change the type and thus the implemented traits of an iterator, including traits that are specialized on, a priori isn’t at all the fault of any unsoundness of a specialization implementaiton. It’s just bad / unsound library design to have a type like `Zip` that’s covariant, yet can be brought into an (unsound) invalid state by subtype coercions using that covariance. And #85874 is about fixing the bad library design in a non-breaking way.

Or even more easily: Variance and coercions are so implicit that they were just forgotten in the design of the `TrustedRandomAccess` trait safety invariants. They weren’t explicitly allowed or forbidden, but the impl. for `vec::IntoIter` was only sound if they’re forbidden and the use case in `Zip`’s implementation was only sound if they’re allowed, because Zip itself is covariant, which is another woefully implicit fact that can be easily overlooked when reasoning about it.

Reply:
How about something like
```rust
/// Like [`TrustedRandomAccess`] but without any of the requirements / guarantees around
/// coercions to subtypes after `__iterator_get_unchecked` (they aren’t allowed here!), and
/// without the requirement that subtypes / supertypes implement `TrustedRandomAccessNoCoerce`.
///
/// This trait was created in PR #85874 to fix soundness issue #85873 without performance regressions.
/// It is subject to change as we might want to build a more generally useful (for performance
/// optimizations) and more sophisticated trait or trait hierarchy that replaces or extends
/// [`TrustedRandomAccess`] and `TrustedRandomAccessNoCoerce`.
#[doc(hidden)]
#[unstable(feature = "trusted_random_access", issue = "none")]
#[rustc_specialization_trait]
pub unsafe trait TrustedRandomAccessNoCoerce: Sized {
```
?

Reply:
Sounds good.